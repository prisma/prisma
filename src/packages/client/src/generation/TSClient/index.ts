import 'flat-map-polyfill' // unfortunately needed as it's not properly polyfilled in TypeScript

export { MinimalArgsType, ArgsType } from './Args'
export { Enum } from './Enum'
export { JS, TS } from './Generatable'
export { InputField, InputType } from './Input'
export { Model, ModelDelegate } from './Model'
export { OutputField, OutputType } from './Output'
export { SchemaOutputField, SchemaOutputType } from './SchemaOutput'
export { TSClient } from './TSClient'

