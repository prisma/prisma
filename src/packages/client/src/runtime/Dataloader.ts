interface Job {
  resolve: (data: any) => void
  reject: (data: any) => void
  request: any
}

export type DataloaderOptions<T> = {
  singleLoader: (request: T) => Promise<any>
  batchLoader: (request: T[]) => Promise<any[]>
  batchBy: (request: T) => string | null
}

export class Dataloader<T = any> {
  batches: { [key: string]: Job[] }
  private tickActive: boolean = false
  constructor(private options: DataloaderOptions<T>) {
    this.batches = {}
  }
  request(request: T): Promise<any> {
    const hash = this.options.batchBy(request)
    if (!hash) {
      return this.options.singleLoader(request)
    }
    if (!this.batches[hash]) {
      this.batches[hash] = []

      // make sure, that we only tick once at a time
      if (!this.tickActive) {
        this.tickActive = true
        process.nextTick(() => {
          this.dispatchBatches()
          this.tickActive = false
        })
      }
    }

    return new Promise((resolve, reject) => {
      this.batches[hash].push({
        request,
        resolve,
        reject,
      })
    })
  }

  private dispatchBatches() {
    for (const key in this.batches) {
      const batch = this.batches[key]
      delete this.batches[key]

      // only batch if necessary
      // this might occur, if there's e.g. only 1 findUnique in the batch
      if (batch.length === 1) {
        this.options
          .singleLoader(batch[0].request)
          .then((result) => {
            if (result instanceof Error) {
              batch[0].reject(result)
            } else {
              batch[0].resolve(result)
            }
          })
          .catch((e) => {
            batch[0].reject(e)
          })
      } else {
        this.options
          .batchLoader(batch.map((j) => j.request))
          .then((results) => {
            if (results instanceof Error) {
              for (let i = 0; i < batch!.length; i++) {
                batch![i].reject(results)
              }
            } else {
              for (let i = 0; i < batch!.length; i++) {
                const value = results[i]
                if (value instanceof Error) {
                  batch![i].reject(value)
                } else {
                  batch![i].resolve(value)
                }
              }
            }
          })
          .catch((e) => {
            for (let i = 0; i < batch!.length; i++) {
              batch![i].reject(e)
            }
          })
      }
    }
  }
}
