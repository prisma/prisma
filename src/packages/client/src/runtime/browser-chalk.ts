const identity = (str) => str || ''

const browserChalk: any = new Proxy(identity, {
  get: (obj, prop) => {
    return browserChalk
  },
  apply: (target, that, args) => {
    return target(args[0])
  },
})

export default browserChalk
