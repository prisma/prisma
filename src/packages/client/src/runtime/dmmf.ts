import { DMMF } from '@prisma/generator-helper'
import { Dictionary, keyBy, ScalarTypeTable, keyBy2 } from './utils/common'

export class DMMFClass implements DMMF.Document {
  public datamodel: DMMF.Datamodel
  public schema: DMMF.Schema
  public mappings: DMMF.Mapping[]
  public queryType: DMMF.OutputType
  public mutationType: DMMF.OutputType
  public outputTypes: DMMF.OutputType[]
  public outputTypeMap: Dictionary<DMMF.OutputType> = {}
  public inputTypes: DMMF.InputType[]
  public inputTypeMap: Dictionary<DMMF.InputType>
  public enumMap: Dictionary<DMMF.SchemaEnum>
  public modelMap: Dictionary<DMMF.Model>
  public mappingsMap: Dictionary<DMMF.Mapping>
  public rootFieldMap: Dictionary<DMMF.SchemaField>
  constructor({ datamodel, schema, mappings }: DMMF.Document) {
    this.datamodel = datamodel
    this.schema = schema
    this.mappings = mappings
    this.enumMap = this.getEnumMap()
    this.queryType = this.getQueryType()
    this.mutationType = this.getMutationType()
    this.modelMap = this.getModelMap()

    this.outputTypes = this.getOutputTypes()

    this.outputTypeMap = this.getMergedOutputTypeMap()

    this.resolveOutputTypes(this.outputTypes)

    this.inputTypes = this.schema.inputTypes
    this.inputTypeMap = this.getInputTypeMap()
    this.resolveInputTypes(this.inputTypes)
    this.resolveFieldArgumentTypes(this.outputTypes, this.inputTypeMap)
    this.mappingsMap = this.getMappingsMap()

    // needed as references are not kept
    this.queryType = this.outputTypeMap.Query
    this.mutationType = this.outputTypeMap.Mutation
    this.outputTypes = this.outputTypes
    this.rootFieldMap = this.getRootFieldMap()
  }
  protected outputTypeToMergedOutputType = (
    outputType: DMMF.OutputType,
  ): DMMF.OutputType => {
    const model = this.modelMap[outputType.name]
    return {
      ...outputType,
      isEmbedded: model ? model.isEmbedded : false,
      fields: outputType.fields,
    }
  }
  protected resolveOutputTypes(types: DMMF.OutputType[]) {
    for (const type of types) {
      for (const field of type.fields) {
        if (
          typeof field.outputType.type === 'string' &&
          !ScalarTypeTable[field.outputType.type]
        ) {
          field.outputType.type =
            this.outputTypeMap[field.outputType.type] ||
            this.enumMap[field.outputType.type] ||
            field.outputType.type
        }
      }
      type.fieldMap = keyBy(type.fields, 'name')
    }
  }
  protected resolveInputTypes(types: DMMF.InputType[]) {
    for (const type of types) {
      for (const field of type.fields) {
        const first = field.inputTypes[0].type
        if (
          typeof first === 'string' &&
          !ScalarTypeTable[first] &&
          (this.inputTypeMap[first] || this.enumMap[first])
        ) {
          field.inputTypes[0].type =
            this.inputTypeMap[first] ||
            this.enumMap[first] ||
            field.inputTypes[0].type
        }
        const second = field.inputTypes[1] && field.inputTypes[1].type
        if (
          typeof second === 'string' &&
          !ScalarTypeTable[second] &&
          (this.inputTypeMap[second] || this.enumMap[second])
        ) {
          field.inputTypes[1].type =
            this.inputTypeMap[second] ||
            this.enumMap[second] ||
            field.inputTypes[1].type
        }
      }
      type.fieldMap = keyBy(type.fields, 'name')
    }
  }
  protected resolveFieldArgumentTypes(
    types: DMMF.OutputType[],
    inputTypeMap: Dictionary<DMMF.InputType>,
  ) {
    for (const type of types) {
      for (const field of type.fields) {
        for (const arg of field.args) {
          const first = arg.inputTypes[0].type
          if (typeof first === 'string' && !ScalarTypeTable[first]) {
            arg.inputTypes[0].type =
              inputTypeMap[first] ||
              this.enumMap[first] ||
              arg.inputTypes[0].type
          }
          const second = arg.inputTypes[1] && arg.inputTypes[1].type
          if (
            second &&
            typeof second === 'string' &&
            !ScalarTypeTable[second]
          ) {
            arg.inputTypes[1].type =
              inputTypeMap[second] ||
              this.enumMap[second] ||
              arg.inputTypes[1].type
          }
        }
      }
    }
  }
  protected getQueryType(): DMMF.OutputType {
    return this.schema.outputTypes.find((t) => t.name === 'Query')!
  }
  protected getMutationType(): DMMF.OutputType {
    return this.schema.outputTypes.find((t) => t.name === 'Mutation')!
  }
  protected getOutputTypes(): DMMF.OutputType[] {
    return this.schema.outputTypes.map(this.outputTypeToMergedOutputType)
  }
  protected getEnumMap(): Dictionary<DMMF.SchemaEnum> {
    return keyBy(this.schema.enums, 'name')
  }
  protected getModelMap(): Dictionary<DMMF.Model> {
    return keyBy(this.datamodel.models, 'name')
  }
  protected getMergedOutputTypeMap(): Dictionary<DMMF.OutputType> {
    return keyBy(this.outputTypes, 'name')
  }
  protected getInputTypeMap(): Dictionary<DMMF.InputType> {
    return keyBy(this.schema.inputTypes, 'name')
  }
  protected getMappingsMap(): Dictionary<DMMF.Mapping> {
    return keyBy(this.mappings, 'model')
  }
  protected getRootFieldMap(): Dictionary<DMMF.SchemaField> {
    return keyBy2(this.queryType.fields, this.mutationType.fields, 'name')
  }
}
