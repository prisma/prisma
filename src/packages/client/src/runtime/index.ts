export { DMMF } from './dmmf-types'
export { DMMFClass } from './dmmf'
export {
  makeDocument,
  transformDocument,
  unpack,
  PrismaClientValidationError,
} from './query'
export { default as debugLib } from '@prisma/debug'

export {
  Engine,
  PrismaClientKnownRequestError,
  PrismaClientUnknownRequestError,
  PrismaClientInitializationError,
  PrismaClientRustPanicError,
} from '@prisma/engine-core'
export { getPrismaClient, PrismaClientOptions } from './getPrismaClient'

export {
  RawValue,
  Sql,
  Value,
  empty,
  join,
  raw,
  sqltag,
} from 'sql-template-tag'


export { default as Decimal } from 'decimal.js'
