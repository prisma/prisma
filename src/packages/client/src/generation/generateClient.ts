import copy from '@apexearth/copy'
import {
  BinaryPaths,
  DataSource,
  DMMF,
  GeneratorConfig,
} from '@prisma/generator-helper'
import fs from 'fs'
import makeDir from 'make-dir'
import path from 'path'
import chalk from 'chalk'
import { promisify } from 'util'
import { DMMF as PrismaClientDMMF } from '../runtime/dmmf-types'
import { Dictionary } from '../runtime/utils/common'
import { getPrismaClientDMMF } from './getDMMF'
import { resolveDatasources } from '../utils/resolveDatasources'
import { extractSqliteSources } from './extractSqliteSources'
import { TSClient, TS, JS } from './TSClient'
import { getVersion } from '@prisma/sdk/dist/engineCommands'
import pkgUp from 'pkg-up'

const remove = promisify(fs.unlink)
const writeFile = promisify(fs.writeFile)
const exists = promisify(fs.exists)
const copyFile = promisify(fs.copyFile)
const stat = promisify(fs.stat)

export class DenylistError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'DenylistError'
    this.stack = undefined
  }
}

export interface GenerateClientOptions {
  datamodel: string
  datamodelPath: string
  browser?: boolean
  schemaDir?: string
  transpile?: boolean
  runtimePath?: string
  outputDir: string
  generator?: GeneratorConfig
  dmmf: DMMF.Document
  datasources: DataSource[]
  binaryPaths: BinaryPaths
  testMode?: boolean
  copyRuntime?: boolean
  engineVersion: string
  clientVersion: string
}

export interface BuildClientResult {
  fileMap: Dictionary<string>
  prismaClientDmmf: PrismaClientDMMF.Document
}

// eslint-disable-next-line @typescript-eslint/require-await
export async function buildClient({
  datamodel,
  schemaDir = process.cwd(),
  runtimePath = '@prisma/client/runtime',
  browser = false,
  binaryPaths,
  outputDir,
  generator,
  dmmf,
  datasources,
  engineVersion,
  clientVersion,
}: GenerateClientOptions): Promise<BuildClientResult> {
  const document = getPrismaClientDMMF(dmmf)

  const client = new TSClient({
    document,
    runtimePath,
    browser,
    datasources: resolveDatasources(datasources, schemaDir, outputDir),
    sqliteDatasourceOverrides: extractSqliteSources(
      datamodel,
      schemaDir,
      outputDir,
    ),
    generator,
    platforms: Object.keys(binaryPaths.queryEngine!),
    schemaDir,
    outputDir,
    clientVersion,
    engineVersion,
  })

  const fileMap = {
    'index.d.ts': TS(client),
    'index.js': JS(client),
  }

  return {
    fileMap,
    prismaClientDmmf: document,
  }
}

async function getDotPrismaDir(outputDir: string): Promise<string> {
  if (
    process.env.INIT_CWD &&
    process.env.npm_lifecycle_event === 'postinstall' &&
    !process.env.PWD?.includes('.pnpm')
  ) {
    // INIT_CWD is the dir, in which "npm install" has been invoked. That can e.g. be in ./src
    // If we're in ./ - there'll also be a package.json, so we can directly go for it
    // otherwise, we'll go up in the filesystem and look for the first package.json
    if (fs.existsSync(path.join(process.env.INIT_CWD, 'package.json'))) {
      return path.join(process.env.INIT_CWD, 'node_modules/.prisma/client')
    }
    const packagePath = await pkgUp({ cwd: process.env.INIT_CWD })
    if (packagePath) {
      return path.join(path.dirname(packagePath), 'node_modules/.prisma/client')
    }
  }

  return path.join(outputDir, '../../.prisma/client')
}

export async function generateClient({
  datamodel,
  datamodelPath,
  schemaDir = datamodelPath ? path.dirname(datamodelPath) : process.cwd(),
  outputDir,
  transpile,
  runtimePath,
  browser,
  generator,
  dmmf,
  datasources,
  binaryPaths,
  testMode,
  copyRuntime,
  clientVersion,
  engineVersion,
}: GenerateClientOptions): Promise<BuildClientResult | undefined> {
  const useDotPrisma = testMode ? !runtimePath : !generator?.isCustomOutput

  runtimePath =
    runtimePath || (useDotPrisma ? '@prisma/client/runtime' : './runtime')

  const finalOutputDir = useDotPrisma
    ? await getDotPrismaDir(outputDir)
    : outputDir

  const { prismaClientDmmf, fileMap } = await buildClient({
    datamodel,
    datamodelPath,
    schemaDir,
    transpile,
    runtimePath,
    browser,
    outputDir: finalOutputDir,
    generator,
    dmmf,
    datasources,
    binaryPaths,
    clientVersion,
    engineVersion,
  })

  const denylistsErrors = validateDmmfAgainstDenylists(
    prismaClientDmmf,
    generator?.previewFeatures,
  )

  if (denylistsErrors) {
    let message = `${chalk.redBright.bold(
      'Error: ',
    )}The schema at "${datamodelPath}" contains reserved keywords.\n       Rename the following items:`

    for (const error of denylistsErrors) {
      message += '\n         - ' + error.message
    }

    message += `\nTo learn more about how to rename models, check out https://pris.ly/d/naming-models`

    throw new DenylistError(message)
  }

  await makeDir(finalOutputDir)
  await makeDir(path.join(outputDir, 'runtime'))

  await Promise.all(
    Object.entries(fileMap).map(async ([fileName, file]) => {
      const filePath = path.join(finalOutputDir, fileName)
      // The deletion of the file is necessary, so VSCode
      // picks up the changes.
      if (await exists(filePath)) {
        await remove(filePath)
      }
      await writeFile(filePath, file)
    }),
  )
  const runtimeSourceDir = testMode
    ? eval(`require('path').join(__dirname, '../../runtime')`) // tslint:disable-line
    : eval(`require('path').join(__dirname, '../runtime')`) // tslint:disable-line

  // if users use a custom output dir
  if (
    copyRuntime ||
    !path.resolve(outputDir).endsWith(`@prisma${path.sep}client`)
  ) {
    // TODO: Windows, / is not working here...
    const copyTarget = path.join(outputDir, 'runtime')
    await makeDir(copyTarget)
    if (runtimeSourceDir !== copyTarget) {
      await copy({
        from: runtimeSourceDir,
        to: copyTarget,
        recursive: true,
        parallelJobs: process.platform === 'win32' ? 1 : 20,
        overwrite: true,
      })
    }
  }

  if (!binaryPaths.queryEngine) {
    throw new Error(
      `Prisma Client needs \`queryEngine\` in the \`binaryPaths\` object.`,
    )
  }

  if (transpile) {
    for (const filePath of Object.values(binaryPaths.queryEngine)) {
      const fileName = path.basename(filePath)
      const target = path.join(finalOutputDir, fileName)
      const [sourceFileSize, targetFileSize] = await Promise.all([
        fileSize(filePath),
        fileSize(target),
      ])

      // If the target doesn't exist yet, copy it
      if (!targetFileSize) {
        await copyFile(filePath, target)
        continue
      }

      // If target !== source size, they're definitely different, copy it
      if (
        targetFileSize &&
        sourceFileSize &&
        targetFileSize !== sourceFileSize
      ) {
        await copyFile(filePath, target)
        continue
      }

      // They must have an equal size now, let's check for the hash
      const [sourceVersion, targetVersion] = await Promise.all([
        getVersion(filePath).catch(() => null),
        getVersion(target).catch(() => null),
      ])

      if (sourceVersion && targetVersion && sourceVersion === targetVersion) {
        // skip
      } else {
        await copyFile(filePath, target)
      }
    }
  }

  const datamodelTargetPath = path.join(finalOutputDir, 'schema.prisma')
  if (datamodelPath !== datamodelTargetPath) {
    await copyFile(datamodelPath, datamodelTargetPath)
  }

  const packageJsonTargetPath = path.join(finalOutputDir, 'package.json')
  const pkgJson = JSON.stringify(
    {
      name: '.prisma/client',
      main: 'index.js',
      types: 'index.d.ts',
    },
    null,
    2,
  )
  await writeFile(packageJsonTargetPath, pkgJson)

  if (process.env.INIT_CWD) {
    const backupPath = path.join(
      process.env.INIT_CWD,
      'node_modules/.prisma/client',
    )
    if (finalOutputDir !== backupPath && !generator?.isCustomOutput) {
      await copy({
        from: finalOutputDir,
        to: backupPath,
        recursive: true,
        parallelJobs: process.platform === 'win32' ? 1 : 20,
        overwrite: true,
      })
    }
  }

  if (transpile) {
    await writeFile(path.join(outputDir, 'runtime/index.d.ts'), backup)
  }

  const proxyIndexJsPath = path.join(outputDir, 'index.js')
  const proxyIndexDTSPath = path.join(outputDir, 'index.d.ts')
  if (!fs.existsSync(proxyIndexJsPath)) {
    await copyFile(path.join(__dirname, '../../index.js'), proxyIndexJsPath)
  }

  if (!fs.existsSync(proxyIndexDTSPath)) {
    await copyFile(path.join(__dirname, '../../index.d.ts'), proxyIndexDTSPath)
  }

  return { prismaClientDmmf, fileMap }
}

const backup = `/// <reference types="node" />

export { DMMF } from './dmmf-types'
import { inspect } from "util";
export declare type Value = string | number | boolean | object | null | undefined;
export declare type RawValue = Value | Sql;

/**
 * A SQL instance can be nested within each other to build SQL strings.
 */
export declare class Sql {
    values: Value[];
    strings: string[];
    rawStrings: ReadonlyArray<string>;
    rawValues: ReadonlyArray<RawValue>;
    constructor(rawStrings: ReadonlyArray<string>, rawValues: ReadonlyArray<RawValue>);
    readonly text: string;
    readonly sql: string;
    [inspect.custom](): {
        text: string;
        sql: string;
        values: Value[];
    };
}
/**
 * Create a SQL query for a list of values.
 */
export declare function join(values: RawValue[], separator?: string): Sql;
/**
 * Create raw SQL statement.
 */
export declare function raw(value: string): Sql;
/**
 * Placeholder value for "no text".
 */
export declare const empty: Sql;
/**
 * Create a SQL object from a template string.
 */
export declare function sqltag(strings: TemplateStringsArray, ...values: RawValue[]): Sql;
/**
 * Standard \`sql\` tag.
 */


export declare var Engine: any
export declare type Engine = any

// export declare var DMMF: any
// export declare type DMMF = any

export declare var DMMFClass: any
export declare type DMMFClass = any

export declare var deepGet: any
export declare type deepGet = any

export declare var chalk: any
export declare type chalk = any

export declare var deepSet: any
export declare type deepSet = any

export declare var makeDocument: any
export declare type makeDocument = any

export declare var transformDocument: any
export declare type transformDocument = any

export declare var debug: any
export declare type debug = any

export declare var debugLib: any
export declare type debugLib = any

export declare var InternalDatasource: any
export declare type InternalDatasource = any

export declare var Datasource: any
export declare type Datasource = any

export declare var printDatasources: any
export declare type printDatasources = any

export declare var printStack: any
export declare type printStack = any

export declare var mergeBy: any
export declare type mergeBy = any

export declare var unpack: any
export declare type unpack = any

export declare var getDMMF: any
export declare type getDMMF = any

export declare var stripAnsi: any
export declare type stripAnsi = any

export declare var parseDotenv: any
export declare type parseDotenv = any

export declare var sqlTemplateTag: any
export declare type sqlTemplateTag = any

export declare class PrismaClientKnownRequestError extends Error {
  code: string;
  meta?: object;
  constructor(message: string, code: string, meta?: any);
}

export declare class PrismaClientUnknownRequestError extends Error {
  constructor(message: string);
}

export declare class PrismaClientRustPanicError extends Error {
    constructor(message: string);
}

export declare class PrismaClientInitializationError extends Error {
    constructor(message: string);
}

export declare class PrismaClientValidationError extends Error {
    constructor(message: string);
}
`

async function fileSize(name: string): Promise<number | null> {
  try {
    const statResult = await stat(name)
    return statResult.size
  } catch (e) {
    return null
  }
}

function validateDmmfAgainstDenylists(
  prismaClientDmmf: PrismaClientDMMF.Document,
  previewFeatures: string[] = [],
): Error[] | null {
  const errorArray = [] as Error[]

  const denylists = {
    // A copy of this list is also in prisma-engines. Any edit should be done in both places.
    // https://github.com/prisma/prisma-engines/blob/master/libs/datamodel/core/src/transform/ast_to_dml/reserved_model_names.rs
    models: [
      // reserved GraphQL words
      'Query',
      'Mutation',
      'Subscription',
      'dmmf',
      'PromiseType',
      'PromiseReturnType',
      'Enumerable',
      'MergeTruthyValues',
      'CleanupNever',
      'Subset',
      'AtLeastOne',
      'atMostOne',
      'OnlyOne',
      'StringFilter',
      'IDFilter',
      'FloatFilter',
      'IntFilter',
      'BooleanFilter',
      'DateTimeFilter',
      'NullableStringFilter',
      'NullableIDFilter',
      'NullableFloatFilter',
      'NullableIntFilter',
      'NullableBooleanFilter',
      'NullableDateTimeFilter',
      'PrismaClientFetcher',
      'PrismaClient',
      'Engine',
      'BatchPayload',
      'Datasources',
      'ErrorFormat',
      'Hooks',
      'LogLevel',
      'LogDefinition',
      'GetLogType',
      'GetEvents',
      'QueryEvent',
      'LogEvent',
      'ModelDelegate',
      'QueryDelegate',
      'missingArg',
      'ArgError',
      'InvalidFieldError',
      'InvalidFieldNameError',
      'InvalidFieldTypeError',
      'EmptySelectError',
      'NoTrueSelectError',
      'IncludeAndSelectError',
      'EmptyIncludeError',
      'InvalidArgError',
      'InvalidArgNameError',
      'MissingArgError',
      'InvalidArgTypeError',
      'AtLeastOneError',
      'AtMostOneError',
      'PrismaClientRequestError',
      'PrismaClientOptions',
      'PrismaClientKnownRequestError',
      'PrismaClientUnknownRequestError',
      'PrismaClientInitializationError',
      'PrismaClientRustPanicError',
      'PrismaVersion',
      'QueryMode',
      'SortOrder',
      'NestedIntFilter',
      'NestedIntNullableFilter',
      'IntNullableFilter',
      'NestedFloatFilter',
      'NestedFloatNullableFilter',
      'FloatNullableFilter',
      'NestedBoolFilter',
      'BoolFilter',
      'NestedBoolNullableFilter',
      'BoolNullableFilter',
      'NestedStringFilter',
      'NestedStringNullableFilter',
      'StringNullableFilter',
      'NestedDateTimeFilter',
      'NestedDateTimeNullableFilter',
      'DateTimeNullableFilter',
      'NestedJsonFilter',
      'JsonFilter',
      'NestedJsonNullableFilter',
      'JsonNullableFilter',
      'InputJsonValue',
      'IntFieldUpdateOperationsInput',
      'FloatFieldUpdateOperationsInput',
      'BoolFieldUpdateOperationsInput',
      'StringFieldUpdateOperationsInput',
      'DateTimeFieldUpdateOperationsInput',
      'JsonFieldUpdateOperationsInput',
      'NullableIntFieldUpdateOperationsInput',
      'NullableFloatFieldUpdateOperationsInput',
      'NullableBoolFieldUpdateOperationsInput',
      'NullableStringFieldUpdateOperationsInput',
      'NullableDateTimeFieldUpdateOperationsInput',
      'NullableJsonFieldUpdateOperationsInput',
      'NullableFloatFieldUpdateOperationsInput',
      // JavaScript keywords
      'await',
      'async',
      'break',
      'case',
      'catch',
      'class',
      'const',
      'continue',
      'debugger',
      'default',
      'delete',
      'do',
      'else',
      'enum',
      'export',
      'extends',
      'false',
      'finally',
      'for',
      'function',
      'if',
      'implements',
      'import',
      'in',
      'instanceof',
      'interface',
      'let',
      'new',
      'null',
      'package',
      'private',
      'protected',
      'public',
      'return',
      'super',
      'switch',
      'this',
      'throw',
      'true',
      'try',
      'typeof',
      'var',
      'void',
      'while',
      'with',
      'yield',
    ],
    fields: ['AND', 'OR', 'NOT'],
    dynamic: [] as string[],
  }

  for (const { name } of prismaClientDmmf.datamodel.models) {
    denylists.dynamic.push(
      ...[
        `${name}Select`,
        `${name}Include`,
        `${name}Default`,
        `${name}Delegate`,
        `${name}GetPayload`,
        `${name}Filter`,

        `${name}Args`,
        `${name}ArgsFilter`,
        `${name}ArgsRequired`,

        `${name}WhereInput`,
        `${name}WhereUniqueInput`,
        `${name}CreateInput`,
        `${name}UpdateInput`,
        `${name}UpdateManyMutationInput`,
        `${name}OrderByInput`,

        `${name}CreateArgs`,

        `${name}UpsertArgs`,

        `${name}UpdateArgs`,
        `${name}UpdateManyArgs`,

        `${name}DeleteArgs`,
        `${name}DeleteManyArgs`,
        `Extract${name}SelectDeleteArgs`,
        `Extract${name}IncludeDeleteArgs`,

        `FindOne${name}Args`,

        `FindMany${name}Args`,

        /** Aggregate Types */

        `Aggregate${name}`,
        `${name}AvgAggregateOutputType`,
        `${name}SumAggregateOutputType`,
        `${name}MinAggregateOutputType`,
        `${name}MaxAggregateOutputType`,
        `${name}AvgAggregateInputType`,
        `${name}SumAggregateInputType`,
        `${name}MinAggregateInputType`,
        `${name}MaxAggregateInputType`,
        `Aggregate${name}Args`,
        `Get${name}AggregateType`,
        `Get${name}AggregateScalarType`,
      ],
    )
  }

  if (prismaClientDmmf.datamodel.enums) {
    for (const it of prismaClientDmmf.datamodel.enums) {
      if (
        denylists.models.includes(it.name) ||
        denylists.fields.includes(it.name) ||
        denylists.dynamic.includes(it.name)
      ) {
        errorArray.push(Error(`"enum ${it.name}"`))
      }
    }
  }

  if (prismaClientDmmf.datamodel.models) {
    for (const it of prismaClientDmmf.datamodel.models) {
      if (
        denylists.models.includes(it.name) ||
        denylists.fields.includes(it.name) ||
        denylists.dynamic.includes(it.name)
      ) {
        errorArray.push(Error(`"model ${it.name}"`))
      }
    }
  }

  return errorArray.length > 0 ? errorArray : null
}
