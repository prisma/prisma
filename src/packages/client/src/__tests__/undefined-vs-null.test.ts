import stripAnsi from 'strip-ansi'
import { blog } from '../fixtures/blog'
import { DMMFClass } from '../runtime/dmmf'
import { makeDocument } from '../runtime/query'
import { getDMMF } from '../generation/getDMMF'

let dmmf
beforeAll(async () => {
  const dmmfDocument = await getDMMF({ datamodel: blog })
  dmmf = new DMMFClass(dmmfDocument)
})

describe('select validation', () => {
  test('null when undefined is allowed', () => {
    const select = {
      data: {
        id: null,
      },
      where: {
        id: 'abc',
      },
    }

    const document = makeDocument({
      dmmf,
      select,
      rootTypeName: 'mutation',
      rootField: 'updateOnePost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(select)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`

        Invalid \`prisma.updateOnePost()\` invocation:

        {
          data: {
        ?   id?: String | StringFieldUpdateOperationsInput,
        ?   createdAt?: DateTime | DateTimeFieldUpdateOperationsInput,
        ?   updatedAt?: DateTime | DateTimeFieldUpdateOperationsInput,
        ?   published?: Boolean | BoolFieldUpdateOperationsInput,
        ?   title?: String | StringFieldUpdateOperationsInput,
        ?   content?: String | NullableStringFieldUpdateOperationsInput | null,
        ?   optionnal?: Float | NullableFloatFieldUpdateOperationsInput | null,
        ?   author?: {
        ?     create?: UserCreateWithoutPostsInput,
        ?     connect?: UserWhereUniqueInput,
        ?     disconnect?: Boolean,
        ?     delete?: Boolean,
        ?     update?: UserUpdateWithoutPostsInput,
        ?     upsert?: UserUpsertWithoutPostsInput,
        ?     connectOrCreate?: UserCreateOrConnectWithoutpostsInput
        ?   },
        ?   categories?: {
        ?     create?: CategoryCreateWithoutPostsInput | CategoryCreateWithoutPostsInput,
        ?     connect?: CategoryWhereUniqueInput | CategoryWhereUniqueInput,
        ?     set?: CategoryWhereUniqueInput | CategoryWhereUniqueInput,
        ?     disconnect?: CategoryWhereUniqueInput | CategoryWhereUniqueInput,
        ?     delete?: CategoryWhereUniqueInput | CategoryWhereUniqueInput,
        ?     update?: CategoryUpdateWithWhereUniqueWithoutPostsInput | CategoryUpdateWithWhereUniqueWithoutPostsInput,
        ?     updateMany?: CategoryUpdateManyWithWhereWithoutPostsInput | CategoryUpdateManyWithWhereWithoutPostsInput,
        ?     deleteMany?: CategoryScalarWhereInput | CategoryScalarWhereInput,
        ?     upsert?: CategoryUpsertWithWhereUniqueWithoutPostsInput | CategoryUpsertWithWhereUniqueWithoutPostsInput,
        ?     connectOrCreate?: CategoryCreateOrConnectWithoutpostsInput | CategoryCreateOrConnectWithoutpostsInput
        ?   }
          },
          where: {
            id: 'abc'
          }
        }

        Argument data.id of type StringFieldUpdateOperationsInput needs at least one argument. Available args are listed in green.

        Note: Lines with ? are optional.

      `)
    }
  })
  test('null when undefined is not allowed', () => {
    const select = {
      data: {
        published: true,
        title: null,
      },
    }

    const document = makeDocument({
      dmmf,
      select,
      rootTypeName: 'mutation',
      rootField: 'createOnePost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(select)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`

                                        Invalid \`prisma.createOnePost()\` invocation:

                                        {
                                          data: {
                                            published: true,
                                            title: null
                                                   ~~~~
                                          }
                                        }

                                        Argument title: Got invalid value null on prisma.createOnePost. Provided null, expected String.


                              `)
    }
  })
})
