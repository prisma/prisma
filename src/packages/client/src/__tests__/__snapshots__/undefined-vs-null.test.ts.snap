// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select validation null when undefined is allowed 1`] = `
"mutation {
  updateOnePost(
    data: {
      id: null
    }
    where: {
      id: \\"abc\\"
    }
  ) {
    id
    createdAt
    updatedAt
    published
    title
    content
    authorId
    optionnal
  }
}"
`;

exports[`select validation null when undefined is not allowed 1`] = `
"mutation {
  createOnePost(data: {
    published: true
    title: null
  }) {
    id
    createdAt
    updatedAt
    published
    title
    content
    authorId
    optionnal
  }
}"
`;
